### PR TITLE
Refactor jukebox sal to use models

### PR DESCRIPTION
**Pull request by AhmedHanafy725**
### Description

- Refactor jukebox sal to use models.
- Clean up the code.
- Add a delete button to the nodes in the deployment.
- Remove ssh key from deployment chatflow.
- Add update deployment instance service.
- Add self-healing service.
- Locking deployment model while writing.
- Fix listing by deployment type.
- Delete deployment in case of failure.

### Related Issues
#10
#45
#46
#47
#48
#49

### CODE REVIEW

File: jumpscale/packages/jukebox/chats/digibyte.py [hunk 1]
```diff
@@ -24,14 +23,12 @@ class DigibyteDeploy(JukeboxDeployChatflow):
 
     @chatflow_step(title="User configurations")
     def environment(self):
-        self.env = {
-            "pub_key": self.public_key,
-            }
+        self.env = {}
+        # self.env = {
+        #     "pub_key": self.public_key,
+        # }
```
Comment (waleedhammam) [inline jumpscale/packages/jukebox/chats/digibyte.py:29]:
```suggestion
```

File: jumpscale/packages/jukebox/chats/ubuntu.py [hunk 2]
```diff
@@ -25,13 +24,10 @@ class UbuntuDeploy(JukeboxDeployChatflow):
 
     @chatflow_step(title="User configurations")
     def environment(self):
-        self.env = {"pub_key": self.public_key}
+        self.env = {}
+        # self.env = {"pub_key": self.public_key}
```
Comment (waleedhammam) [inline jumpscale/packages/jukebox/chats/ubuntu.py:28]:
```suggestion
```

File: jumpscale/packages/jukebox/serivces/monitor.py [hunk 3]
```diff
@@ -16,42 +16,46 @@ def __init__(self, interval=60 * 60 * 12, *args, **kwargs):
...
+            gevent.sleep(1)
 
         j.logger.info("All deployments are monitored")
 
-    def _check_down_containers_and_auto_extend_pools(self, all_deployments, identity_name):
+    def _check_down_containers(self, deployment):
+        user = j.data.text.removeprefix(deployment.identity_name, "jukebox_")
+
+        # for deployment_type, deployments in all_deployments.items():
+        #     for deployment in deployments:
```
Comment (waleedhammam) [inline jumpscale/packages/jukebox/serivces/monitor.py:32]:
```suggestion
```

File: jumpscale/packages/jukebox/serivces/monitor.py [hunk 4]
```diff
@@ -16,42 +16,46 @@ def __init__(self, interval=60 * 60 * 12, *args, **kwargs):
...
+
+        number_of_deployed = len([node for node in deployment.nodes if node.state == State.DEPLOYED])
+        number_of_down_deployment = deployment.nodes_count - number_of_deployed
+        if number_of_down_deployment > 0:
+            j.logger.warning(
+                f"Deployment: {deployment.deployment_name}, Deployment type: {deployment.solution_type}, owner: {user} has {number_of_down_deployment} node(s) went down"
+            )
+            message = (
+                f"Dear {user},\n\n"
+                f"{number_of_down_deployment} node(s) of {deployment.solution_type} went down for your deployment {deployment.deployment_name}."
```
Comment (waleedhammam) [inline jumpscale/packages/jukebox/serivces/monitor.py:42]:
```suggestion
                f"{number_of_down_deployment} node(s) of {deployment.solution_type} went down for your deployment {deployment.deployment_name}. The system will try to bring it back, Please refer for your jukebox dashboard for more information."
```

File: jumpscale/sals/jukebox/__init__.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+    def cleanup(self, deployment):
+        for blockchain_node in deployment.nodes:
+            deployment.zos.workloads.decomission(blockchain_node.wid)
+            blockchain_node.state = State.DELETED
+            deployment.save()
+
+
+BCNodeFACTORY = BlockchainStoredFactory(JukeboxDeployment)
+BCNodeFACTORY.always_reload = True
```
Comment (waleedhammam) [inline jumpscale/sals/jukebox/__init__.py:62]: better to be at the top of the file ?

File: jumpscale/sals/jukebox/jukebox.py [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+            network_view.name,
+            node,
+            self.pool_ids[0],
+            network_view_copy,
+            identity_name=self.identity_name,
+            owner=self.identity_name,
+        )
+
+        if result:
+            # self.md_show_update("Deploying Network on Nodes....")
```
Comment (waleedhammam) [inline jumpscale/sals/jukebox/jukebox.py:104]:
```suggestion
```

File: jumpscale/sals/jukebox/jukebox.py [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+            network_name=network_name,
+            ip_address=ip_address,
+            flist=flist,
+            cpu=self.cpu,
+            memory=self.memory,
+            disk_size=self.disk_size,
+            env=env,
+            interactive=False,
+            entrypoint=entry_point,
+            # log_config=self.log_config,
```
Comment (waleedhammam) [inline jumpscale/sals/jukebox/jukebox.py:222]:
```suggestion
```

File: jumpscale/sals/jukebox/jukebox_deploy_chatflow.py [hunk 8]
```diff
@@ -1,31 +1,48 @@
...
         "get_deployment_name",
         "block_chain_info",
         "choose_farm",
         "set_expiration",
-        "upload_public_key",
         "payment",
         "deploy",
         "success",
     ]
     # FLIST = "<flist_url>"
```
Comment (waleedhammam) [inline jumpscale/sals/jukebox/jukebox_deploy_chatflow.py:44]:
```suggestion
```

File: jumpscale/sals/jukebox/utils.py [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+
+def get_possible_farms(cru, sru, mru, number_of_deployments):
+    gcc = GlobalCapacityChecker()
+    farm_names = gcc.get_available_farms(
+        cru=cru * number_of_deployments,
+        mru=mru * number_of_deployments,
+        sru=sru * number_of_deployments,
+        # ip_version=None,
+        # no_nodes=1,
```
Comment (waleedhammam) [inline jumpscale/sals/jukebox/utils.py:70]:
```suggestion
```

**Review by waleedhammam — APPROVED** [review]

File: jumpscale/sals/jukebox/jukebox_deploy_chatflow.py (hunk 8 above)
Comment (AhmedHanafy725) [reply jumpscale/sals/jukebox/jukebox_deploy_chatflow.py:44]: We need this for anyone adding a new child chatflow

File: jumpscale/sals/jukebox/__init__.py (hunk 5 above)
Comment (AhmedHanafy725) [reply jumpscale/sals/jukebox/__init__.py:62]: no, it needs to be here after the class definition

**Review by waleedhammam — APPROVED** [review]

**Pull request merged**